RFC-0001-Agile Coretime.md

```diff
@@ -0,0 +1,265 @@
+
+### Broker-chain
+
+The Broker-chain is a new system parachain. It has the responsibility of providing the Relay-chain via UMP with scheduling information of:
+
+- How many cores which should be made available during the next session.
+- Which cores should be allocated to which para IDs.
+
+Any cores left unallocated are assumed by the Broker-chain to be used for Instantaneous Coretime sales.
+
+It is also expected to receive information from the Relay-chain on total revenue from instantaneous coretime sales on a per-block basis.
+
+### Bulk Sales
+
+There is a periodic sale of Coretime happening at a period of `BULK_PERIOD`.  A number of instances of Coretime spanning from `LEADIN_PERIOD` in the future to `LEADIN_PERIOD + BULK_PERIOD` in the future of a single Polkadot UC Core is offered by the Polkadot System at a fixed price.
+
+These instances which are owned by purchaser are called *regions*. This sale happens on the Broker-chain. Regions are quantized into atomic periods of `TIMESLICE`, into which `BULK_PERIOD` divides a whole number of times. The `Timeslice` type is a `u16` which can be multiplied by `TIMESLICE` to give a `BlockNumber` value indicating the same period in times of (Relay-chain) blocks.
+
+The Broker-chain aims to sell some `BULK_TARGET` of Cores, up to some `BULK_LIMIT`. It makes this sale in a single batch `LEADIN_PERIOD` prior to the beginning of the period being sold. The Broker chain publishes a price for this batch of sales for the `BULK_TARGET` period prior to the sale execution.
+
+Accounts may call a `purchase` function with the appropriate funds in place to have their funds reserved and signal an intention to purchase Bulk Coretime for the forthcoming period. One account may have only one pending purchase. Any number of accounts may attempt to `purchase` Bulk Coretime for the forthcoming period, but the order is recorded.
+
+If there are more purchases than available cores for purchase in this period, then any additional purchase orders are carried over but marked as such. A purchase is only cancellable if it was carried over.
+
+When a block of Bulk Coretime is initially issued through this purchase, the price it was purchased for is recorded, in addition to the beginning and end Relay-chain block numbers to which it corresponds.
+
+The Broker-chain SHALL record this information in a storage double-map called Regions, keyed first by the current bulk sale index (a `u32` starting at zero and incrementing with each sale), then secondarily by a `RegionId`. It shall map into a value of `RegionRecord`:
+
+```rust
+struct RegionId {
+    core: CoreIndex, // A `u16`.
+    begin: Timeslice,    // A `u16`.
+}
+struct RegionRecord {
+    owner: AccountId,
+    end: Timeslice,
+    price: Option<Balance>,
+    allocation: Option<Vec<ParaId>>, // begins set to `None`
+}
+```
+
+Notably, if a region is split or transferred, then the `price` is reset to `None`.
```

Comment:
I think it would be easier to understand if you explicitly say that splitting a region removes the ability to renew at the same price which I think you are implicitly saying by specifying splitting sets the price to “none”. This also implies though that “split cores” are not eligible for priority renewal, correct? You also don’t seem to mention anything of the fact that existing owners of cores should be able to get priority renewal. Is that correct?

Reply:
This is written above:

> Notably, if a region is split or transferred, then the `price` is reset to `None`.

Not enough?

Also, this is about split *regions*, i.e. taking the month-long piece and splitting in into smaller pieces.

Reply:
I don’t understand if there is a specific reason that the price needs to be set to none. Does that mean a split region can’t be renewed the same way a “full core” region would?

Reply:
Yeah - because it would then have two owners - which one could "renew"?

Reply:
My comment initially was to clarify that split cores shouldn’t be able to get priority renewal. Carefully reading the spec does imply that but I think making this more explicit would be helpful for people to understand.

> I think it's ok when the core would be used consistently by the same paras in the same way from month to month, but it doesn't make sense when they're being carved up and, presumably, traded.

I think this does not go into the initial spec. It would should be possible to offer to buy split cores at some point at which point this can be added. I doubt the demand for them will be particularly high today and it shouldn’t delay a first implementation.